src/Parsers/PostgreSQL/ParserPostgreSQLQuery.cpp

```diff
@@ -0,0 +1,67 @@
+#include "Parsers/Lexer.h"
+#include "Transform.h"
+#include <Parsers/PostgreSQL/test/TestData/Select.h>
+#include <Parsers/PostgreSQL/ParserPostgreSQLQuery.h>
+#include <Parsers/PostgreSQL/Common/Types.h>
+#include <Parsers/PostgreSQL/Common/util/JSONHelpers.h>
+#include <Parsers/ParserQuery.h>
+#include <Parsers/ParserSetQuery.h>
+#include <Parsers/parseQuery.h>
+#include <Parsers/iostream_debug_helpers.h>
+
+#include <pg_query.h>
+
+#include <Common/logger_useful.h>
+#include <Poco/Logger.h>
+
+#include <string>
+
+#include "config.h"
+
+#include <base/scope_guard.h>
+#include <iostream>
+
+namespace DB
+{
+
+bool ParserPostgreSQLQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
+{
+    ParserSetQuery set_p;
+
+    if (set_p.parse(pos, node, expected))
+        return true;
+
+    const auto * begin = pos->begin;
+
+    // The same parsers are used in the client and the server, so the parser have to detect the end of a single query in case of multiquery queries
+    while (!pos->isEnd() && pos->type != TokenType::Semicolon)
+        ++pos;
+
+    const auto * end = pos->begin;
+
+    String query(begin, end);
+
+    std::cerr << "Query: " << query << std::endl;
+
+    auto res = pg_query_parse(query.data());
+
+    String json = String(res.parse_tree);
```

Comment:
It is pointless to obtain the AST in JSON and then parse it back.
We probably could work on the tree representation directly to avoid this serialization/deserialization step.